[ARM] Fix musttail calls

ostannard opened the pull request:
This is a continuation of #102896 by @kiran-isaac, which gets the `[[clang::musttail]]` attribute working for as many cases as possible in the ARM backend.

This includes one target-independent LangRef change to the `musttail` IR attribute, which I think is a case which can't work for any target, but wasn't previously documented.

With this patch series applied, I'm still aware of these bugs/limitations in musttail for ARM:
* Tail calls aren't supported at all for Armv6-M, because it doesn't have a branch with enough range that does not modify LR. 
* With M-profile PACBTI, there are some indirect calls (ones which have arguments in all of r0-r3) where no register is available to hold the function pointer, so these can't be tail-called.
* Some crash bugs which I've raised separate issues for:
  * #107569
  * #109922
  * #109929

ostannard committed:
[ARM] Re-generate a test

ostannard committed:
[ARM] Add debug trace for tail-call optimisation

There are lots of reasons a call might not be eligible for tail-call
optimisation, this adds debug trace to help understand the compiler's
decisions here.

ostannard committed:
[ARM] Tail-calls do not require caller and callee arguments to match

The ARM backend was checking that the outgoing values for a tail-call
matched the incoming argument values of the caller. This isn't
necessary, because the caller can change the values in both registers
and the stack before doing the tail-call. The actual limitation is that
the callee can't need more stack space for it's arguments than the
caller does.

This is needed for code using the musttail attribute, as well as
enabling tail calls as an optimisation in more cases.

kiran-isaac committed:
[Clang] Always forward sret parameters to musttail calls

If a call using the musttail attribute returns it's value through an
sret argument pointer, we must forward an incoming sret pointer to it,
instead of creating a new alloca. This is always possible because the
musttail attribute requires the caller and callee to have the same
argument and return types.

ostannard committed:
[ARM] Fix comment typo

ostannard committed:
[ARM] Allow tail calls with byval args

Byval arguments which are passed partially in registers get stored into
the local stack frame, but it is valid to tail-call them because the
part which gets spilled is always re-loaded into registers before doing
the tail-call, so it's OK for the spill area to be deallocated.

ostannard committed:
[ARM] Allow functions with sret returns to be tail-called

It is valid to tail-call a function which returns through an sret
argument, as long as we have an incoming sret pointer to pass on.

efriedma-quic commented on llvm/docs/LangRef.rst:
```diff
@@ -12658,10 +12658,10 @@ This instruction requires several arguments:
-   varargs from the caller. Calls marked ``musttail`` must obey the following
-   additional  rules:
+   Both markers imply that the callee does not access allocas or ``byval``
+   arguments from the caller. The ``tail`` marker additionally implies that the
```
This is fine, but please submit separately.

rnk replied:
I think you want to say that this case is not supported:
```
struct ByVal { int large[5]; };
void f(ByVal*);
void g(ByVal o) {
  [[clang::musttail]] f(&o);
}
```
The LangRef needs to permit *forwarding* cases like this, though:
```
struct ByVal { int large[5]; };
void f(ByVal a, ByVal b);
void g(ByVal a, ByVal b) {
  [[clang::musttail]] f(b, a);
}
```

We can make this work by copying the arguments out and back in before tearing down the call frame.

Can you please clarify?

ostannard replied:
That was my intention, but the wording here didn't allow it. I've split the langref change out into #110093, and updated the wording. I'll leave the old patch in here for now to avoid a force-push.

efriedma-quic commented on llvm/test/CodeGen/ARM/musttail.ll:
```diff
@@ -0,0 +1,321 @@
+  store i32 5, ptr %4, align 4
+  musttail call void @large_callee(%large_struct* byval(%large_struct) align 4 %y)
+  ret void
+}
```
https://github.com/llvm/llvm-project/pull/102896#discussion_r1739257241

ostannard replied:
Good catch, we generate wrong code for that case, because we try do to the copies in-place without temporary storage. I'll see if there's a reasonable way to fix that, and if not I might have to back out the byval change for now.

efriedma-quic commented on clang/lib/CodeGen/CGCall.cpp:
```diff
@@ -5112,7 +5112,7 @@ RValue CodeGenFunction::EmitCall(const CGFunctionInfo &CallInfo,
   llvm::Value *UnusedReturnSizePtr = nullptr;
   if (RetAI.isIndirect() || RetAI.isInAlloca() || RetAI.isCoerceAndExpand()) {
-    if (IsVirtualFunctionPointerThunk && RetAI.isIndirect()) {
+    if ((IsVirtualFunctionPointerThunk && RetAI.isIndirect()) || IsMustTail) {
```
https://github.com/llvm/llvm-project/pull/104795#discussion_r1722204520

ostannard replied:
I think `isIndirect()` is the only case needed for `IsMustTail`, so I'll change the code to limit it to that, matching the `IsVirtualFunctionPointerThunk` check.

rnk reviewed (commented):
Thanks for prioritizing this! I have triaged many internal issues about this.

ostannard committed:
[ARM] Avoid clobbering byval arguments when passing to tail-calls

When passing byval arguments to tail-calls, we need to store them into
the stack memory in which this the caller received it's arguments. If
any of the outgoing arguments are forwarded from incoming byval
arguments, then the source of the copy is from the same stack memory.
This can result in the copy corrupting a value which is still to be
read.

The fix is to first make a copy of the outgoing byval arguments in local
stack space, and then copy them to their final location. This fixes the
correctness issue, but results in extra copying, which could be
optimised.

ostannard committed:
[ARM] Optimise byval arguments in tail-calls

We don't need to copy byval arguments to tail calls via a temporary, if
we can prove that we are not copying from the outgoing argument area.
This patch does this when the source if the argument is one of:
* Memory in the local stack frame, which can't be used for tail-call
  arguments.
* A global variable.

We can also avoid doing the copy completely if the source and
destination are the same memory location, which is the case when the
caller and callee have the same signature, and pass some arguments
through unmodified.

rnk commented:
By the way, the x86 backend also miscompiles test cases like this:
```
struct ByVal { int large[5]; };
void f(ByVal a, ByVal b);
void g(ByVal a, ByVal b) {
  [[clang::musttail]] f(b, a);
}
```

I have an internal issue assigned to @aeubanks tracking that. I may have reported it upstream. It should be looked into. It uses the analogous copy-pasted "MatchingStackOffset" helpers which have been removed in this change, and any technique used for ARM is probably applicable to x86.

I think we'd get decent code by doing a sequence of `LOAD <all outgoing byval bytes to vregs> ; TOKENFACTOR ; STORE <all vregs to byval argument slots>`. This would allow the register allocator to use the register file as temporary storage to implement the swap, instead of having to conservatively allocate temporary stack space to copy every outgoing byval arguement, or implement complicated aliasing rules to avoid the need to do that.

efriedma-quic commented:
> I think we'd get decent code by doing a sequence of LOAD <all outgoing byval bytes to vregs> ; TOKENFACTOR ; STORE <all vregs to byval argument slots>. 

Up to some threshold, sure.  If the scheduler is sufficiently clever, we won't even get any spills: we can interleave the loads and stores.  But we probably still need some fallback path that avoids emitting large numbers of inline load/store ops.

ostannard commented:
@rnk I've already written a patch which modifies the existing strategy of using pseudo-instructions to do memory-memory copies, adding in a temporary on the stack where needed.

efriedma-quic commented on llvm/test/CodeGen/ARM/fp-arg-shuffle.ll:
```diff
@@ -1,8 +1,31 @@
 ; RUN: llc -mtriple=arm-eabi -mattr=+neon -float-abi=soft %s -o - | FileCheck %s
 
 ; CHECK: function1
 ; CHECK-NOT: vmov
```
Clean up old CHECK-NOT?

ostannard replied:
Done

efriedma-quic commented on llvm/test/CodeGen/ARM/musttail.ll:
```diff
@@ -0,0 +1,343 @@
+; CHECK:       @ %bb.0: @ %entry
+; CHECK-NEXT:    .pad #16
+; CHECK-NEXT:    sub sp, sp, #16
+; CHECK-NEXT:    stm sp!, {r0, r1, r2, r3}
```
This store is unnecessary (the stack memory is deallocated immediately after the values are stored).  Not a big deal, but maybe leave a TODO.

ostannard replied:
Done

ostannard commented:
Rebased over #110093, and squashed the fixup commits.

stefson commented:
so, it seems I've hit one of these in compiling firefox at src/gl.cc

ostannard commented:
The libc++ test failures look unrelated, rebased onto current main to check.

ostannard commented:
That rebase picked up a test failure in `CodeGen/AArch64/sme-callee-save-restore-pairs.ll` which looks related to this but isn't, I'll rebase again over the fix.

ostannard commented:
Ping.

efriedma-quic commented on llvm/test/CodeGen/ARM/musttail.ll:
```diff
@@ -0,0 +1,345 @@
+  ret void
+}
+
+; TODO: This test swaps the two byval arguments, but does so without copying to
```
Is this TODO still relevant?

ostannard replied:
No, I've removed it.

ostannard committed:
fixup! [ARM] Avoid clobbering byval arguments when passing to tail-calls

Remove old comment

ostannard commented:
Ping

statham-arm reviewed (approved):
LGTM. Especially convenient to have all the patches split apart from each other like that – it made them much easier to read!
